src/main/java/org/springframework/retry/support/MetricsRetryListener.java

```diff
@@ -0,0 +1,132 @@
+ * limitations under the License.
+ */
+
+package org.springframework.retry.support;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.Timer;
+
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.RetryListener;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link RetryListener} implementation for Micrometer {@link Timer}s around retry
+ * operations.
+ * <p>
+ * The {@link Timer#start} is called from the {@link #open(RetryContext, RetryCallback)}
+ * and stopped in the {@link #close(RetryContext, RetryCallback, Throwable)}. This
+ * {@link Timer.Sample} is associated with the provided {@link RetryCallback} to make this
+ * {@link MetricsRetryListener} instance reusable for many retry operation.
+ * <p>
+ * The registered {@value #TIMER_NAME} {@link Timer} has these tags by default:
+ * <ul>
+ * <li>{@code name} - {@link RetryCallback#getLabel()}</li>
+ * <li>{@code retry.count} - the number of attempts - 1; essentially the successful first
+ * call means no counts</li>
+ * <li>{@code exception} - the thrown back to the caller (after all the retry attempts)
+ * exception class name</li>
+ * </ul>
+ * <p>
+ * The {@link #setCustomTags(Map)} and {@link #setCustomTagsProvider(Function)} can be
+ * used to further customize tags on the timers.
+ *
+ * @author Artem Bilan
+ * @since 2.0.8
+ */
+public class MetricsRetryListener implements RetryListener {
+
+	public static final String TIMER_NAME = "spring.retry";
+
+	private final MeterRegistry meterRegistry;
+
+	private final Map<RetryCallback<?, ?>, Timer.Sample> retryContextToSample = new HashMap<>();
+
+	private final Map<String, String> customTags = new HashMap<>();
+
+	private final Meter.MeterProvider<Timer> retryMeterProvider;
+
+	private Function<RetryContext, Iterable<Tag>> customTagsProvider = retryContext -> Tags.empty();
+
+	/**
+	 * Construct an instance based on the provided {@link MeterRegistry}.
+	 * @param meterRegistry the {@link MeterRegistry} to use for timers.
+	 */
+	public MetricsRetryListener(MeterRegistry meterRegistry) {
+		Assert.notNull(meterRegistry, "'meterRegistry' must not be null");
+		this.meterRegistry = meterRegistry;
+		this.retryMeterProvider = Timer.builder(TIMER_NAME)
+			.description("Metrics for Spring RetryTemplate")
+			.withRegistry(this.meterRegistry);
+	}
+
+	/**
+	 * Supply a map of tags which are going to be used for all the timers managed by this
+	 * listener.
+	 * @param customTags the map of additional tags for all the timers.
+	 */
+	public void setCustomTags(Map<String, String> customTags) {
+		Assert.notNull(customTags, "'customTags' must not be null");
+		this.customTags.putAll(customTags);
+	}
```

Comment:
I would do this in the constructor instead, also If you use `Tags` instead of `Map` for `customTags` above, you can do this (`Tags` implements `Iterable<Tag>` ):
```suggestion
	public void setCustomTags(Iterable<Tag> customTags) {
		Assert.notNull(customTags, "'customTags' must not be null");
		this.customTags.and(customTags);
	}
```

Reply:
I'm not sure what you mean with constructor, but since this is optional feature, then our style is to have it as setter.
But I agree that these custom tags can go to the `retryMeterProvider` immediately from here instead of later on in the `close`.

Reply:
Well, no: the `MeterProvider` is not a builder and all of its methods returns for us a result `Meter`. 
So, taking that back. 
Now I see why you have insisted on moving it to the ctor! 😄